File: frontend/src/types/recommendation.ts
/**
 * Recommendation API contract (Frontend 06 / FE6-00).
 *
 * Aligns with Backend recommendation draft on `origin/feature/backend/policy-recommendation`.
 * W4-G0 proposal — not approved as current public API contract.
 *
 * @see docs/development/develop_plan/integration/06_recommendation_vertical_slice.md
 */

import type { PolicyCategory } from './policy.js';

/** Client router path — distinct from NL search at `/search`. */
export const RECOMMENDATION_APP_ROUTE = '/recommendations';

export const RECOMMENDATION_ENDPOINTS = {
  post: {
    method: 'POST' as const,
    path: '/api/v1/recommendations',
  },
  get: {
    method: 'GET' as const,
    path: '/api/v1/policies/recommendations',
  },
} as const;

/** Backend draft status filter (`open` | `upcoming` | `closed`). Differs from public `ApplicationStatus`. */
export type RecommendationStatusFilter = 'open' | 'upcoming' | 'closed';

export interface RecommendationRequest {
  age?: number | null;
  region?: string | null;
  /** Backend accepts string category code (e.g. finance, housing). */
  category?: PolicyCategory | string | null;
  status?: RecommendationStatusFilter | string | null;
  include_partial?: boolean;
  limit?: number;
}

export interface RecommendationReasonDto {
  code: string;
  label: string;
}

export interface RecommendationItemDto {
  id: number;
  source_id: string;
  external_id: string;
  title: string;
  lead: string | null;
  category: string;
  regions: string[];
  min_age: number | null;
  max_age: number | null;
  application_start: string | null;
  application_end: string | null;
  application_status: string;
  data_quality_status: string;
  /** Request-internal ordering only — not eligibility probability (W4-G0). */
  score: number;
  reasons: RecommendationReasonDto[];
  unknown_conditions: string[];
  disclaimer: string;
}

export interface RecommendationResponse {
  items: RecommendationItemDto[];
  total: number;
  evaluated_at: string;
}

export const RECOMMENDATION_DEFAULTS = {
  include_partial: false,
  limit: 10,
} as const;

export const RECOMMENDATION_LIMITS = {
  ageMin: 0,
  ageMax: 120,
  limitMin: 1,
  limitMax: 50,
} as const;

export const RECOMMENDATION_DEFAULT_DISCLAIMER =
  '본 추천 결과는 자격을 확정하지 않으며, 상세 자격 및 신청 조건은 공식 원문에서 확인해야 합니다.';

export interface ResolvedRecommendationRequest {
  age?: number;
  region?: string;
  category?: string;
  status?: string;
  include_partial: boolean;
  limit: number;
}

export interface RecommendationValidationErrorBody {
  detail: string;
}

export function resolveRecommendationRequest(
  request: RecommendationRequest = {},
): ResolvedRecommendationRequest {
  const include_partial = request.include_partial ?? RECOMMENDATION_DEFAULTS.include_partial;
  const limit = request.limit ?? RECOMMENDATION_DEFAULTS.limit;

  if (
    request.age !== undefined &&
    request.age !== null &&
    (!Number.isInteger(request.age) ||
      request.age < RECOMMENDATION_LIMITS.ageMin ||
      request.age > RECOMMENDATION_LIMITS.ageMax)
  ) {
    throw new Error('Recommendation age must be an integer from 0 to 120.');
  }

  if (
    !Number.isInteger(limit) ||
    limit < RECOMMENDATION_LIMITS.limitMin ||
    limit > RECOMMENDATION_LIMITS.limitMax
  ) {
    throw new Error('Recommendation limit must be an integer from 1 to 50.');
  }

  const region = request.region?.trim();
  const category = request.category?.toString().trim();
  const status = request.status?.toString().trim();

  return {
    include_partial,
    limit,
    ...(request.age !== undefined && request.age !== null ? { age: request.age } : {}),
    ...(region ? { region } : {}),
    ...(category ? { category } : {}),
    ...(status ? { status } : {}),
  };
}

File: backend/app/schemas/recommendation.py
from datetime import datetime
from typing import List, Literal, Optional
from pydantic import BaseModel, Field, ConfigDict


RecommendationStatus = Literal["open", "upcoming", "closed"]


class RecommendationRequest(BaseModel):
    """맞춤 정책 추천 요청 DTO."""
    age: Optional[int] = Field(default=None, ge=0, le=120, description="사용자 만 연령")
    region: Optional[str] = Field(default=None, description="거주지 (예: 서울특별시, 경기도)")
    category: Optional[str] = Field(default=None, description="관심 정책 분야 (예: finance, housing, employment, education)")
    categories: List[str] = Field(
        default_factory=list,
        max_length=7,
        description="복수 관심 정책 분야. 하나 이상 부합하면 추천 후보로 평가",
    )
    status: Optional[RecommendationStatus] = Field(
        default=None,
        description="신청 상태 필터 (open, upcoming, closed)",
    )
    include_partial: bool = Field(default=False, description="품질 상태가 partial인 정책 포함 여부")
    limit: int = Field(default=10, ge=1, le=50, description="최대 추천 반환 수")


class RecommendationReason(BaseModel):
    """추천 사유 항목 DTO."""
    code: str = Field(description="추천 사유 코드 (MATCHED_CATEGORY, MATCHED_REGION, MATCHED_AGE, MATCHED_STATUS 등)")
    label: str = Field(description="사람이 읽을 수 있는 추천 사유 설명 문구")


class RecommendationItem(BaseModel):
    """추천 정책 아이템 DTO."""
    id: int
    source_id: str
    external_id: str
    title: str
    lead: Optional[str] = None
    category: str
    regions: List[str] = Field(default_factory=list)
    min_age: Optional[int] = None
    max_age: Optional[int] = None
    application_start: Optional[str] = None
    application_end: Optional[str] = None
    application_status: str
    data_quality_status: str = "valid"
    score: int = Field(description="추천 부합도 점수")
    reasons: List[RecommendationReason] = Field(default_factory=list, description="추천 사유 목록")
    unknown_conditions: List[str] = Field(default_factory=list, description="데이터로 확인할 수 없는 미확정 조건")
    disclaimer: str = Field(
        default="본 추천 결과는 자격을 확정하지 않으며, 상세 자격 및 신청 조건은 공식 원문에서 확인해야 합니다.",
        description="비단정 안내 문구"
    )

    model_config = ConfigDict(from_attributes=True)


class RecommendationResponse(BaseModel):
    """맞춤 정책 추천 응답 DTO."""
    items: List[RecommendationItem]
    total: int
    evaluated_at: str = Field(description="추천 평가 수행 ISO-8601 시각")
